src/query/storage/m3/multi_fetch_result.go

```diff
@@ -117,55 +143,84 @@ func (r *multiResult) Add(
 		return
 	}
 
-	r.seenIters = append(r.seenIters, iterators)
+	if len(r.seenIters) == 0 {
+		// store the first attributes seen
+		r.seenFirstAttrs = attrs
+	}
+	r.seenIters = append(r.seenIters, newIterators)
+
+	// Need to check the error to bail early after accumulating the iterators
+	// otherwise when we close the the multi fetch result
 	if !r.err.Empty() {
 		// don't need to do anything if the final result is going to be an error
 		return
 	}
 
-	iters := iterators.Iters()
-	if r.dedupeMap == nil {
-		r.dedupeMap = make(map[string][]multiResultSeries, len(iters))
+	if len(r.seenIters) < 2 {
+		// don't need to create the de-dupe map until we need to actually need to
+		// dedupe between two results
+		return
 	}
 
-	for _, iter := range iters {
-		id := iter.ID().String()
-		r.dedupeMap[id] = append(r.dedupeMap[id], multiResultSeries{
-			attrs: attrs,
-			iter:  iter,
-		})
+	if len(r.seenIters) == 2 {
+		// need to backfill the dedupe map from the first result first
+		existing := r.seenIters[0]
+		r.dedupeMap = make(map[string]multiResultSeries, existing.Len())
+		for _, iter := range existing.Iters() {
+			id := iter.ID().String()
+			r.dedupeMap[id] = multiResultSeries{
+				attrs: r.seenFirstAttrs,
+				iter:  iter,
+			}
+		}
 	}
 
-	r.dedupe()
+	r.dedupe(attrs, newIterators)
 }
 
-func (r *multiResult) dedupe() {
-	for id, serieses := range r.dedupeMap {
-		if len(serieses) < 2 {
-			continue
-		}
+func (r *multiResult) dedupe(
+	attrs storage.Attributes,
+	newIterators encoding.SeriesIterators,
+) {
+	for _, iter := range newIterators.Iters() {
+		id := iter.ID().String()
 
-		// find max resolution
-		maxIdx := 0
-		maxValue := serieses[0].attrs
-		for i := 1; i < len(serieses); i++ {
-			series := serieses[i].attrs
-			if series.Resolution > maxValue.Resolution {
-				maxIdx = i
-				maxValue = series
+		existing, exists := r.dedupeMap[id]
+		if !exists {
+			// Does not exist, new addition
+			r.dedupeMap[id] = multiResultSeries{
+				attrs: attrs,
+				iter:  iter,
 			}
+			continue
 		}
 
-		maxSeries := serieses[maxIdx]
-
-		// reset slice to reuse
-		for idx := range serieses {
-			serieses[idx] = multiResultSeries{}
+		var existsBetter bool
+		switch r.fanout {
+		case namespaceCoversAllQueryRange:
+			// Already exists and resolution of result we are adding is not as precise
+			existsBetter = existing.attrs.Resolution <= attrs.Resolution
+		case namespaceCoversPartialQueryRange:
+			// Already exists and either has longer retention, or the same retention
+			// and result we are adding is not as precise
+			existsLongerRetention := existing.attrs.Retention > attrs.Retention
```

Comment:
All this is to avoid consolidation... :sigh:

Reply:
Yup =]